File: week-10/pushedrumex/흩날리는 시험지 속에서 내 평점이 느껴진거야.py
N, K = map(int, input().split())
arr = list(map(int, input().split()))

answer = 0
left, right = 0, sum(arr)

while left <= right:
    mid = (left + right) // 2

    # mid 가 답이 되도록 그룹을 나눠보자
    group, score = 0, 0
    for n in arr:
        score += n
        # mid 가 답이 되려면 최솟값 >= mid 이어야함
        if score >= mid:
            group += 1
            score = 0
    # 가능성 있는 점수라면
    if group >= K:
        answer = mid
        left = mid + 1
    else:
        right = mid - 1

print(answer)

File: week-10/dahyen0o/흩날리는-시험지-속에서-내-평점이-느껴진거야.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static StringBuilder sb;

    public static void main(String[] args) throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        sb = new StringBuilder();

        st = new StringTokenizer(br.readLine());
        final int N = Integer.parseInt(st.nextToken());
        final int K = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        final int[] scores = new int[N];
        int MAX_SCORE = 0;
        for (int n = 0; n < scores.length; n++) {
            scores[n] = Integer.parseInt(st.nextToken());
            MAX_SCORE += scores[n];
        }

        int left = 0, right = MAX_SCORE;
        int answer = 0;

        while (left <= right) {
            int minScore = (left + right) / 2;
            
            // K개의 그룹으로 쪼갰을 때 합이 모두 minScore 넘을 수 있을지 검사
            if (canLowerBound(minScore, scores, K)) {
                answer = Math.max(answer, minScore);
                left = minScore + 1;
            } else {
                right = minScore - 1;
            }
        }
        
        bw.write(sb.append(answer).append("\n").toString());
        bw.flush();
    }

    private static boolean canLowerBound(final int lowerBound, final int[] scores, final int K) {
        int k = 0;
        int score = 0;

        for (int i = 0; i < scores.length; i++) {
            score += scores[i];

            if (score >= lowerBound) {
                score = 0;
                k++;
            }
        }

        return k >= K;
    }
}
